File: python-selenium/pages/accountpage.py
# -*- coding: utf-8 -*-

from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC

class AccountPage:
    def __init__(self, webdriver):
        self.webdriver = webdriver

    def click_button_SignUp(self):
        xpath = "//button[text()='Sign Up']";
        wait = WebDriverWait(self.webdriver, 10)
        wait.until(EC.presence_of_element_located((By.XPATH, xpath)))
        wait = WebDriverWait(self.webdriver, 10)
        wait.until(EC.visibility_of_element_located((By.XPATH, xpath)))
        element = self.webdriver.find_element_by_xpath(xpath)
        element.click()

    def findElement_Account_headerText(self):
        # Account_headerTextDesktop_1IzPi
        #
        xpath ="//div[starts-with(@class, 'Account_headerText')]";
        wait = WebDriverWait(self.webdriver, 10)
        wait.until(EC.presence_of_element_located((By.XPATH, xpath)))
        element = self.webdriver.find_element_by_xpath(xpath)
        return element

    def send_input_signupFirstName(self, firstname):
        element = self.webdriver.find_element_by_name("signupFirstName")
        element.send_keys(firstname)

    def send_input_signupLastName(self, lastname):
        element = self.webdriver.find_element_by_name("signupLastName")
        element.send_keys(lastname)

    def send_input_signupEmail(self, email):
        element = self.webdriver.find_element_by_name("signupEmail")
        element.send_keys(email)

    def send_input_signupPassword(self, password):
        element = self.webdriver.find_element_by_name("signupPassword")
        element.send_keys(password)

    def select_Gender(self, gender):

        gender_dropdown_xpath = "//div[starts-with(@class, 'Account_signupBirthdayGenderContainer_')]/div/div/div/div/a"

        gender_dropdown = self.webdriver.find_element_by_xpath(gender_dropdown_xpath)
        gender_dropdown.click();

        options_xpath = "//div[starts-with(@class, 'Account_signupBirthdayGenderContainer_')]/div/div/div[2]/a/div"
        gender_dropdown_elements = self.webdriver.find_elements_by_xpath(options_xpath)
        gender_options = {}
        for gender_dropdown_element in gender_dropdown_elements:
            print(gender_dropdown_element.text + " -> " + str(gender_dropdown_element))
            gender_options[gender_dropdown_element.text] = gender_dropdown_element
        gender_options[gender].click()

    def send_input_birthdate(self, birthdate):
        xpath= "//input[@placeholder='Birthdate']"
        element = self.webdriver.find_element_by_xpath(xpath)
        element.send_keys(birthdate)

    def click_button_CreateProfile(self):
        xpath = "//button[text()='Create Profile']"
        wait = WebDriverWait(self.webdriver, 10)
        wait.until(EC.presence_of_element_located((By.XPATH, xpath)))
        wait = WebDriverWait(self.webdriver, 10)
        wait.until(EC.visibility_of_element_located((By.XPATH, xpath)))
        element = self.webdriver.find_element_by_xpath(xpath)
        element.click()

    def get_element_Account_signupSuccessHeaderText(self):
        xpath = "//div[starts-with(@class, 'Account_signupSuccessHeaderText')]"
        element = self.webdriver.find_element_by_xpath(xpath)
        return element

    def get_element_Account_signupSuccessText(self):
        xpath = "//div[starts-with(@class, 'Account_signupSuccessText')]"
        element = self.webdriver.find_element_by_xpath(xpath)
        return element

    def get_element_Account_signupErrorGenericText(self):
        xpath = "//div[starts-with(@class, 'Account_signupErrorGenericText_')]"
        element = self.webdriver.find_element_by_xpath(xpath)
        return element

    def click_button_Done(self):
        xpath = "//button[text()='Done']"
        wait = WebDriverWait(self.webdriver, 10)
        wait.until(EC.presence_of_element_located((By.XPATH, xpath)))
        wait = WebDriverWait(self.webdriver, 10)
        wait.until(EC.visibility_of_element_located((By.XPATH, xpath)))
        element = self.webdriver.find_element_by_xpath(xpath)
        element.click()

File: java-selenium/src/test/java/com/fox/automation/page/AccountPage.java
package com.fox.automation.page;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.util.HashMap;
import java.util.List;

public class AccountPage {
    private WebDriver driver;

    public AccountPage(WebDriver driver) {
        this.driver = driver;
    }

    public void click_button_SignUp() {
        By By_SignUpButton = By.xpath("//button[text()='Sign Up']");
        WebDriverWait wait = new WebDriverWait(driver, 10);
        wait.until(
                ExpectedConditions.and(
                        ExpectedConditions.presenceOfElementLocated(By_SignUpButton),
                        ExpectedConditions.visibilityOfElementLocated(By_SignUpButton)
                )
        );
        WebElement webElement = driver.findElement(By_SignUpButton);
        webElement.click();

    }

    public WebElement findElement_Account_headerText() {
        /**
         *
         * Find WebElement for Account_textHeader
         *
         * Note: class is different for different sizes:
         *  * Account_headerTextDesktop_
         *  * Account_headerText_
         *
         */
        By By_Account_headerText = By.xpath("//div[starts-with(@class, 'Account_headerText')]");
        WebDriverWait wait = new WebDriverWait(driver, 10);
        wait.until(ExpectedConditions.presenceOfElementLocated(By_Account_headerText));
        List<WebElement> elements =  driver.findElements(By_Account_headerText);
        return elements.get(1);
    }

    public WebElement find_Account_container() {
        String xpath = "//div[@class]";
        WebDriverWait wait = new WebDriverWait(driver, 10);
        wait.until(
                ExpectedConditions.and(
                        ExpectedConditions.presenceOfElementLocated(By.xpath(xpath)),
                        ExpectedConditions.visibilityOfElementLocated(By.xpath(xpath))
                )
        );
        WebElement webElement = driver.findElement(By.xpath(xpath));
        return webElement;
    }

    public void send_input_signupFirstName(String firstname) {
        WebElement webElement=driver.findElement(By.name("signupFirstName"));
        webElement.sendKeys(firstname);
    }

    public void send_input_signupLastName(String firstname) {
        WebElement webElement=driver.findElement(By.name("signupLastName"));
        webElement.sendKeys(firstname);
    }

    public void send_input_signupEmail(String firstname) {
        WebElement webElement=driver.findElement(By.name("signupEmail"));
        webElement.sendKeys(firstname);
    }

    public void send_input_signupPassword(String firstname) {
        WebElement webElement=driver.findElement(By.name("signupPassword"));
        webElement.sendKeys(firstname);
    }

    public void select_Gender(String gender) {
        /**
         * Here is where things get tricky - the 'Gender' is a dynamic/javascript dropdown - click it once to reveal options
         */

        //String xpath ="//a[text()='Gender']";
        String dropdown_xpath = "//div[starts-with(@class, 'Account_signupBirthdayGenderContainer_')]/div/div/div/div/a";

        WebElement webElement = driver.findElement(By.xpath(dropdown_xpath));
        webElement.click();

        /**
         * Now find/iterate through options to build HashMap
         */
        String options_xpath = "//div[starts-with(@class, 'Account_signupBirthdayGenderContainer_')]/div/div/div[2]/a/div";
        List<WebElement> elements = driver.findElements(By.xpath(options_xpath));

        HashMap<String, WebElement> genderOptions = new HashMap<String, WebElement>();

        for (WebElement element : elements) {
            // System.out.println("gender_selection:" + element.getText() + " -> " + element);
            genderOptions.put(element.getText(), element);
        }

        /**
         * Find and select the option
         */
        webElement = genderOptions.get(gender);
        webElement.click();
    }


    public void send_input_birthdate(String birthdate) {
        String xpath= "//input[@placeholder='Birthdate']";
        WebElement webElement = driver.findElement(By.xpath(xpath));
        webElement.sendKeys(birthdate);
    }


    public void click_button_CreateProfile() {
        String xpath = "//button[text()='Create Profile']";
        WebDriverWait wait = new WebDriverWait(driver, 10);
        wait.until(
                ExpectedConditions.and(
                        ExpectedConditions.presenceOfElementLocated(By.xpath(xpath)),
                        ExpectedConditions.visibilityOfElementLocated(By.xpath(xpath))
                )
        );
        WebElement webElement = driver.findElement(By.xpath(xpath));
        webElement.click();
    }

    public WebElement getElement_Account_signupSuccessHeaderText() {
        String xpath = "//div[starts-with(@class, 'Account_signupSuccessHeaderText')]";
        WebElement webElement = driver.findElement(By.xpath(xpath));
        return webElement;
    }

    public WebElement getElement_Account_signupSuccessText() {
        String xpath = "//div[starts-with(@class, 'Account_signupSuccessText')]";
        WebElement webElement = driver.findElement(By.xpath(xpath));
        return webElement;
    }

    public WebElement getElement_Account_signupErrorGenericText() {
        String xpath = "//div[starts-with(@class, 'Account_signupErrorGenericText_')]";
        WebElement webElement = driver.findElement(By.xpath(xpath));
        return webElement;

    }

    public void click_button_Done() {
        String xpath = "//button[text()='Done']";
        WebDriverWait wait = new WebDriverWait(driver, 10);
        wait.until(
                ExpectedConditions.and(
                        ExpectedConditions.presenceOfElementLocated(By.xpath(xpath)),
                        ExpectedConditions.visibilityOfElementLocated(By.xpath(xpath))
                )
        );
        WebElement webElement = driver.findElement(By.xpath(xpath));
        webElement.click();
    }
}
